sycl/test-e2e/ESIMD/api/mask_math.cpp

```diff
@@ -0,0 +1,164 @@
+//==---------------- mask_math.cpp  - DPC++ ESIMD on-device test -----==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// This test checks basic arithmetic operations between simd and simd_mask
+//===----------------------------------------------------------------------===//
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include "../esimd_test_utils.hpp"
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+namespace esimd = sycl::ext::intel::esimd;
+// Result type of a scalar binary Op
+constexpr unsigned Size = 2;
+template <class T1, class T2, class OpClass>
+using scalar_comp_t =
+    std::conditional_t<std::is_same_v<OpClass, esimd_test::CmpOp>,
+                       typename esimd::simd_mask<Size>::element_type,
+                       __ESIMD_DNS::computation_type_t<T1, T2>>;
+// Number of test cases for every operator type
+constexpr int TestCaseNumber = 7;
+
+// Result type of a vector binary Op
+template <class T1, class T2, class OpClass, int N = 0>
+using comp_t = std::conditional_t<
+    N == 0, scalar_comp_t<T1, T2, OpClass>,
+    std::conditional_t<
+        std::is_same_v<OpClass, esimd_test::CmpOp>, esimd::simd_mask<N>,
+        esimd::simd<__ESIMD_DNS::computation_type_t<T1, T2>, N>>>;
+
+template <class T, class OpClass, class Ops> bool test(Ops ops, queue &q) {
+  T *Input = sycl::malloc_shared<T>(Size, q);
+  constexpr int NumOps = (int)Ops::size;
+  int CSize = NumOps * Size;
+  T *Output = sycl::malloc_shared<T>(CSize * TestCaseNumber, q);
+  for (int i = 0; i < Size; ++i)
+    Input[i] = i;
+  constexpr T TestValue = 2;
+  std::vector<int> expectedValues;
+
+  // esimd_test::ArithBinaryOpsNoDiv has 3 members for general arithmetic
+  // operations: +, -, *.
+  if constexpr (std::is_same_v<OpClass, esimd_test::BinaryOp> &&
+                Ops::size == 3) {
+    expectedValues = {3, 2, 1, 2, 2, 0, 3,  2,  -1, -2, 2,  0,  3, 2,
+                      1, 2, 2, 0, 3, 2, -1, -2, 2,  0,  2,  0,  0, 0,
+                      1, 0, 3, 2, 1, 2, 2,  0,  3,  2,  -1, -2, 2, 0};
+  } else if constexpr (std::is_same_v<OpClass, esimd_test::CmpOp>) {
+    expectedValues = {0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
+                      0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
+                      1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
+                      1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
+                      1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0};
+  } else {
+    // esimd_test::IntBinaryOpsNoShiftNoDivRem has 6 members for integer
+    // arithmetic operations: +, -, *, |, & ^

```

Comment:
This PR adds support for SHIFT, DIV and REM operations. They need to be tested as well.